src/compiler/program.ts

```diff
@@ -2882,6 +2882,7 @@ export function createProgram(rootNamesOrOptions: readonly string[] | CreateProg
     }
 
     function getDeclarationDiagnostics(sourceFile?: SourceFile, cancellationToken?: CancellationToken): readonly DiagnosticWithLocation[] {
+        if (sourceFile && isSourceOfProjectReferenceRedirect(sourceFile.fileName)) return emptyArray;

```

Comment:
Actually i think getDeclarationDiagnostics need to filter files with "sourceFileMayBeEmitted"

Reply:
`sourceFileMayBeEmitted` makes sense here, will need to pass in an appropriate host, though. It also probably makes sense to do the filtering down in the `getDeclarationDiagnostics` core implementation in `declarations.ts`.